public/app/features/browse-dashboards/state/reducers.ts

```diff
@@ -0,0 +1,101 @@
+import { PayloadAction } from '@reduxjs/toolkit';
+
+import { DashboardViewItem, DashboardViewItemKind } from 'app/features/search/types';
+
+import { BrowseDashboardsState } from '../types';
+
+import { fetchChildren } from './actions';
+
+type FetchChildrenAction = ReturnType<typeof fetchChildren.fulfilled>;
+
+export function extraReducerFetchChildrenFulfilled(state: BrowseDashboardsState, action: FetchChildrenAction) {
+  const parentUID = action.meta.arg;
+  const children = action.payload;
+
+  if (!parentUID) {
+    state.rootItems = children;
+    return;
+  }
+
+  state.childrenByParentUID[parentUID] = children;
+
+  // If the parent of the items we've loaded are selected, we must select all these items also
+  const parentIsSelected = state.selectedItems.folder[parentUID];
+  if (parentIsSelected) {
+    for (const child of children) {
+      state.selectedItems[child.kind][child.uid] = true;
+    }
+  }
+}
+
+export function setFolderOpenState(
+  state: BrowseDashboardsState,
+  action: PayloadAction<{ folderUID: string; isOpen: boolean }>
+) {
+  const { folderUID, isOpen } = action.payload;
+  state.openFolders[folderUID] = isOpen;
+}
+
+export function setItemSelectionState(
+  state: BrowseDashboardsState,
+  action: PayloadAction<{ item: DashboardViewItem; isSelected: boolean }>
+) {
+  const { item, isSelected } = action.payload;
+
+  function markChildren(kind: DashboardViewItemKind, uid: string) {
+    state.selectedItems[kind][uid] = isSelected;
+
+    if (kind !== 'folder') {
+      return;
+    }
+
+    let children = state.childrenByParentUID[uid] ?? [];
+    for (const child of children) {
+      markChildren(child.kind, child.uid);
+    }
+  }
+
+  markChildren(item.kind, item.uid);
+
+  // If we're unselecting an item, unselect all ancestors also
+  if (!isSelected) {
+    let nextParentUID = item.parentUID;
+
+    while (nextParentUID) {
+      const parent = findItem(state.rootItems, state.childrenByParentUID, nextParentUID);
+      if (!parent) {
+        break;
+      }
+
+      state.selectedItems[parent.kind][parent.uid] = false;
+      nextParentUID = parent.parentUID;
+    }
```

Comment:
i was struggling to get my head around this logic. had to play around with it in the UI before it sunk in. agree with @tskarhed that some tests or maybe even some comments to explain **why** we want to unselect here could really help clarify the behaviour 👍

Reply:
Yeah, fair. This is also kind of missing the other piece (which we've deprioritised) which is that the parent checkbox would change to a mixed/indeterminate to reflect that it's children are partially checked.

Reply:
added both unit tests, and some comments explaining this :)

lemme know how that goes - i understand both the feature and code is a little bit tricky to grok